### PR TITLE
Add support for decorators with private elements - Spec mode

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
~~:warning: Please review #9166 first, since this PR will change quite a bit if/when that PR is merged.~~

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

NOTE: This PR only implements spec mode, not loose.

ToDo
- [ ] ~~Hide private elements from class decorators~~ (https://github.com/babel/babel/pull/9154#issuecomment-450528483)
- [ ] Option for https://github.com/tc39/proposal-decorators/issues/167 (different PR)
- [x] More tests?

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
I have not finished yet writing the tests, but apart from that this PR is ready to be reviewed!

cc @littledan

**Comment by littledan** [comment]
Sorry for the lateness, but I am considering proposing restrictions to decorators' interaction with private, see https://github.com/tc39/proposal-decorators/issues/180

**Comment by nicolo-ribaudo** [comment]
Oh ok, I missed that issue

**Comment by nicolo-ribaudo** [comment]
Force-pushed with #9166

File: packages/babel-helpers/src/helpers.js [hunk 1]
```diff
@@ -1048,7 +1048,11 @@ helpers.classPrivateFieldGet = helper("7.0.0-beta.0")`
     if (!privateMap.has(receiver)) {
       throw new TypeError("attempted to get private field on non-instance");
     }
-    return privateMap.get(receiver).value;
+    var descriptor = privateMap.get(receiver);
+    if (descriptor.get) {
+      return descriptor.get.call(receiver);
+    }
+    return descriptor.value;
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.js]: I copied this change from https://github.com/babel/babel/pull/9101

File: packages/babel-helpers/src/helpers.js [hunk 2]
```diff
@@ -1058,13 +1062,17 @@ helpers.classPrivateFieldSet = helper("7.0.0-beta.0")`
...
+    if (descriptor.set) {
+      descriptor.set.call(receiver, value);
+    } else {
+      if (!descriptor.writable) {
+        // This should only throw in strict mode, but class bodies are
+        // always strict and private fields can only be used inside
+        // class bodies.
+        throw new TypeError("attempted to set read only private field");
+      }
+      descriptor.value = value;
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.js]: Also this is from https://github.com/babel/babel/pull/9101

File: packages/babel-helpers/src/helpers.js [hunk 3]
```diff
@@ -1256,6 +1264,52 @@ helpers.decorate = helper("7.1.5")`
...
     var api = {
+      toElementKey(key) {
+        return toPropertyKey(key);
+      },
+
+      // ClassElementEvaluation
+      createElementDescriptor: function(
+        def /*: ElementDefinition */,
+      ) /*: ElementDescriptor */ {
+        var key = this.toElementKey(def.key);
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.js]: Except for this line, this function has just been moved.

File: packages/babel-helpers/src/helpers.js [hunk 4]
```diff
@@ -1639,48 +1695,6 @@ helpers.decorate = helper("7.1.5")`
     return api;
   }
 
-  // ClassElementEvaluation
-  function _createElementDescriptor(
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.js]: Moved above.

**Comment by nicolo-ribaudo** [comment]
@littledan I was thinking that Babel should throw when a class with private elements has a class-level decorator; at least until https://github.com/tc39/proposal-decorators/pull/184 is resolved. WDYT?

**Comment by littledan** [comment]
Yeah, apologies for the delay in resolving that issue. I'm thinking we should remove `elements` for class decorators, and just let them add `extras`. What would this mean for Babel, if you've already released a version with `elements`?

**Comment by nicolo-ribaudo** [comment]
We should remove `elements` behind an option and document it as the standard behavior.
Anyway, I was thinking about throwing (for now) when there are both class decorators and private elements, at least until a final decision is taken.

**Comment by littledan** [comment]
Yes, that sounds like a good plan.

**Comment by xlexi** [comment]
Hello, I just ran into the "Private fields in decorated classes are not supported yet" error and found this PR upon searching for it and I see no activity since January, what is holding it back? Is there anything I can assist with

**Comment by nicolo-ribaudo** [comment]
The decorators proposal is having a lot of changes currently, and it is completely different from the version implemented in Babel.

**Comment by nicolo-ribaudo** [comment]
This doesn't reflect the decorators proposal anymore.

**Pull request closed without merging**